In [1]:
%load_ext google.cloud.bigquery

# "Will it snow tomorrow?" - The time traveler asked
The following dataset contains climate information form over 9000 stations accross the world. The overall goal of these subtasks will be to predict whether it will snow tomorrow 11 years ago. So if today is 2021.02.15 then the weather we want to forecast is for the date 2010.02.16. You are suppsed to solve the tasks using Big Query, which can be used in the Jupyter Notebook like it is shown in the following cell. For further information and how to used BigQuery in Jupyter Notebook refer to the Google Docs. 

The goal of this test is, to test your coding knowledge in Python, BigQuery and Pandas as well as your understanding of Data Science. If you get stuck at the first part, you can use the replacement data provided in the second part

In [2]:
%%bigquery 
SELECT
*,
FROM `bigquery-public-data.samples.gsod`
LIMIT 20 

Downloading: 100%|██████████| 20/20 [00:02<00:00,  7.83rows/s]


station_number  wban_number  year  month  day  mean_temp  \
0            39730        99999  1929     10   20  52.799999   
1            33110        99999  1929     12   18  47.500000   
2            37770        99999  1931      4   24  50.200001   
3           726810        24131  1931      6   23  65.099998   
4           726810        24131  1931      3    2  42.799999   
5           726810        24131  1931      9   17  67.000000   
6           726810        24131  1931      8    7  68.400002   
7           726810        24131  1932      7   14  64.099998   
8           726810        24131  1932     10   23  41.099998   
9           726810        24131  1932      1    5  24.600000   
10          726815        24106  1932      8   27  71.000000   
11          726810        24131  1932      8   20  71.000000   
12          726810        24131  1932      5   21  55.200001   
13          370310        99999  1933     10   17  55.299999   
14          292310        99999  1933     12    1 -11.000000   
15          370310        99999  1933      6   17  62.700001   
16          239330        99999  1933      6    4  70.500000   
17          282750        99999  1933      1    7  -5.000000   
18          292310        99999  1933      3   17 -10.300000   
19          370310        99999  1933      4   23  65.000000   

    num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
0                       4       45.500000                         4.0   
1                       4       44.000000                         4.0   
2                       4       44.299999                         4.0   
3                      24       41.500000                         8.0   
4                      24       31.500000                         8.0   
5                      24       40.500000                         8.0   
6                      24       37.200001                         8.0   
7                      24       54.099998                         8.0   
8                      24       31.000000                         8.0   
9                      24       21.100000                         8.0   
10                     24             NaN                         NaN   
11                     24       41.700001                         8.0   
12                     24       46.599998                         8.0   
13                      4             NaN                         NaN   
14                      4             NaN                         NaN   
15                      4             NaN                         NaN   
16                      4             NaN                         NaN   
17                      4             NaN                         NaN   
18                      4             NaN                         NaN   
19                      4             NaN                         NaN   

    mean_sealevel_pressure  ...  min_temperature  min_temperature_explicit  \
0                      NaN  ...              NaN                      None   
1                      NaN  ...              NaN                      None   
2                      NaN  ...              NaN                      None   
3                      NaN  ...              NaN                      None   
4                      NaN  ...              NaN                      None   
5                      NaN  ...              NaN                      None   
6                      NaN  ...              NaN                      None   
7                      NaN  ...              NaN                      None   
8                      NaN  ...              NaN                      None   
9                      NaN  ...              NaN                      None   
10                     NaN  ...              NaN                      None   
11                     NaN  ...              NaN                      None   
12                     NaN  ...              NaN                      None   
13                     NaN  ...  

## Part 1

In [3]:
import datetime, timedelta
import pandas as pd
from sklearn.model_selection import train_test_split

### 1. Task
Change the date format to 'YYYY-MM-DD' and select the data from 2005 till 2010 for station numbers including and between 725300 and 726300 , and save it as a pandas dataframe. Note the maximum year available is 2010. 

In [4]:
# Select data and save it as pandas dataframe "weather_data_1"

In [5]:
%%bigquery weather_data_1
SELECT
*,
FROM `bigquery-public-data.samples.gsod`
WHERE
    year >= 2005 AND year <= 2010 AND station_number >= 725300 AND station_number <= 726300

Downloading: 100%|██████████| 399713/399713 [00:24<00:00, 16201.96rows/s]


In [6]:
# Create new column with the date in format 'YYYY-MM-DD'
year = weather_data_1['year'].astype(str)
month = weather_data_1['month'].astype(str)
day = weather_data_1['day'].astype(str)
weather_data_1['date'] = pd.to_datetime(year + "-" + month + "-" + day)

In [7]:
weather_data_1.head()

station_number  wban_number  year  month  day  mean_temp  \
0          725835        99999  2005      4    6  60.799999   
1          725786        99999  2005      1    1  27.900000   
2          725868        99999  2005     12    6  17.100000   
3          725869        99999  2005      7    6  76.400002   
4          725869        99999  2005      6   13  55.200001   

   num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
0                      4       27.000000                         4.0   
1                      7       26.900000                         7.0   
2                      8       15.600000                         8.0   
3                      9       48.400002                         9.0   
4                      9       32.400002                         9.0   

   mean_sealevel_pressure  ...  min_temperature_explicit  total_precipitation  \
0                     NaN  ...                      None                 0.00   
1                     NaN  ...                      None                 0.45   
2                     NaN  ...                      None                 0.00   
3                     NaN  ...                      None                 0.00   
4                     NaN  ...                      None                 0.00   

   snow_depth    fog   rain   snow   hail  thunder  tornado       date  
0         NaN  False  False  False  False    False    False 2005-04-06  
1         NaN  False  False  False  False    False    False 2005-01-01  
2         5.1  False  False  False  False    False    False 2005-12-06  
3         NaN  False  False  False  False    False    False 2005-07-06  
4         NaN  False  False  False  False    False    False 2005-06-13  

[5 rows x 32 columns]

### 2. Task 
From here want to work with the data from all stations 725300 to 725330 that have information from 2005 till 2010. 

In [8]:
# In "weather_data_1" are already saved all data from 2005 till 2010 and from stations 725300 to 726300.
# So, we need just reduce the stations to 725330.
weather_data_2 = weather_data_1.loc[weather_data_1['station_number'] <= 725330]

In [9]:
weather_data_2.head()

station_number  wban_number  year  month  day  mean_temp  \
141          725326        99999  2005      3   23  36.799999   
150          725326        99999  2005     11   22  36.299999   
208          725326        99999  2005     12   21   9.400000   
345          725326        99999  2006      1    4  39.000000   
384          725326        99999  2006      8   24  76.099998   

     num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
141                     24       27.200001                        24.0   
150                     24       24.299999                        24.0   
208                     24        4.800000                        24.0   
345                     24       36.000000                        24.0   
384                     24       65.099998                        24.0   

     mean_sealevel_pressure  ...  min_temperature_explicit  \
141                     NaN  ...                      None   
150                     NaN  ...                      None   
208                     NaN  ...                      None   
345                     NaN  ...                      None   
384                     NaN  ...                      None   

     total_precipitation  snow_depth    fog   rain   snow   hail  thunder  \
141                 0.04         NaN  False  False  False  False    False   
150                 0.00         NaN  False  False  False  False    False   
208                 0.00         NaN  False  False  False  False    False   
345                 0.00         NaN  False  False  False  False    False   
384                 0.00         NaN  False  False  False  False    False   

     tornado       date  
141    False 2005-03-23  
150    False 2005-11-22  
208    False 2005-12-21  
345    False 2006-01-04  
384    False 2006-08-24  

[5 rows x 32 columns]

Do a first analysis of the remaining dataset, clean or drop data depending on how you see appropriate. 

In [10]:
shape = weather_data_2.shape
print("Dataframe \"weather_data_2\" has", shape[0], "rows (datasets) and", shape[1], "columns (features). \n")
print("General information about the dateframe and all features: \n")

Dataframe "weather_data_2" has 19189 rows (datasets) and 32 columns (features). 

General information about the dateframe and all features: 



In [11]:
weather_data_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19189 entries, 141 to 399710
Data columns (total 32 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   station_number                      19189 non-null  int64         
 1   wban_number                         19189 non-null  int64         
 2   year                                19189 non-null  int64         
 3   month                               19189 non-null  int64         
 4   day                                 19189 non-null  int64         
 5   mean_temp                           19189 non-null  float64       
 6   num_mean_temp_samples               19189 non-null  int64         
 7   mean_dew_point                      19187 non-null  float64       
 8   num_mean_dew_point_samples          19187 non-null  float64       
 9   mean_sealevel_pressure              17245 non-null  float64       
 10  num_mean_sealevel_p

We see that 6 features are boolean, 1 feature is datetime, 2 features are of type object and 23 features are numeric (17 float and 6 int). 

In [12]:
weather_data_2.describe(include=['object'])

max_temperature_explicit min_temperature_explicit
count                     19187                        0
unique                        2                        0
top                       False                      NaN
freq                      11187                      NaN

In [13]:
weather_data_2['max_temperature_explicit'] = weather_data_2['max_temperature_explicit'].astype('float64')
weather_data_2.max_temperature_explicit.unique()

<ipython-input-13-07f602520d02>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  weather_data_2['max_temperature_explicit'] = weather_data_2['max_temperature_explicit'].astype('float64')


array([ 1.,  0., nan])

Using the info method, it is convenient to quickly look at the gaps in the data, in our case there are some columns with much less non-null-values than 19189.
We can first go drop this columns: 
- mean_station_pressure, 
- num_mean_station_pressure_samples, 
- max_gust_wind_speed,
- min_temperature, 
- min_temperature_explicit,
- snow_depth.

Also we can drop column 'max_temperature_explicit', because it values have type object, and if we try to convert it in expected type float, remain this 3 values: 0.,  1., nan, so we can not use this feature, so that it makes sense.

In [14]:
weather_data_2__without_null = weather_data_2.drop([
    'mean_station_pressure',
    'num_mean_station_pressure_samples',
    'max_gust_wind_speed',
    'max_temperature_explicit',
    'min_temperature',
    'min_temperature_explicit',
    'snow_depth',
], axis = 1)

Values from columns 'year', 'month' and 'day' are already saved in new column 'date', so we can drop them too. 

In [15]:
weather_data_2__without_null__new_date = weather_data_2__without_null.drop([
    'year',
    'month',
    'day',
], axis = 1)

Some columns are not informative or helpfull for our task, e.g. all columns with samples number:
- num_mean_temp_samples,
- num_mean_dew_point_samples,
- num_mean_sealevel_pressure_samples,
- num_mean_visibility_samples,
- num_mean_wind_speed_samples.

In [16]:
weather_data_2__without_null__new_date__without_num_samples = weather_data_2__without_null__new_date.drop([
    'num_mean_temp_samples',
    'num_mean_dew_point_samples',
    'num_mean_sealevel_pressure_samples',
    'num_mean_visibility_samples',
    'num_mean_wind_speed_samples',
], axis = 1)

Some infos to 'station_number' and 'wban_number':

- (station_number?) WMO Identifiers - The World Meteorological Organization relies on a 5-digit numeric code to identify a weather station. It is widely used in synoptic and upper air reports. The entire identifier is often called the "index number". The first two digits are sometimes referred to as the "block number" and refer to the geographic area (00-29 Europe, 30-59 Asia, 60-69 Africa, 70-79 North America, 80-89 South America, 90-99 Oceania). The last three digits are loosely referred to as the "station number" in the context of "block numbers". 
- WBAN - The WBAN (Weather Bureau Army Navy) identifier is a 5-digit identifier developed in the 1950s to augment the system of longline identifiers. It is still used by NCDC to identify many of its climatological datasets and thus continues to be important. 

[http://www.weathergraphics.com/identifiers/index.htm]

We need one of this features as identifier, where exactly will we predict the weather.

In [17]:
station_number = pd.unique(weather_data_2__without_null__new_date__without_num_samples.station_number).tolist()
wban_number = pd.unique(weather_data_2__without_null__new_date__without_num_samples.wban_number).tolist()
station_number_length = len(station_number)
wban_number_length = len(wban_number)
print("Number of station_number's: ", station_number_length)
print("station_number's: ", station_number)
print("Number of wban_number's: ", wban_number_length)
print("wban_number's: ", wban_number)

Number of station_number's:  10
station_number's:  [725326, 725327, 725317, 725316, 725315, 725305, 725320, 725300, 725314, 725330]
Number of wban_number's:  11
wban_number's:  [99999, 4894, 3887, 14842, 94846, 3960, 14827, 94870, 4846, 53802, 94892]


station_number's look more logical (the same prefix indicating the location of stations), so I will use it and drop the another identifier - wban_number.

In [18]:
weather_data_2__without_null__new_date__without_num_samples__station_number = weather_data_2__without_null__new_date__without_num_samples.drop([
    'wban_number',
], axis = 1)

In [19]:
# Drop all rows that have any NaN values
weather_data_2__without_null__new_date__without_num_samples__station_number__without_NaN = weather_data_2__without_null__new_date__without_num_samples__station_number.dropna()

In [20]:
weather_data = weather_data_2__without_null__new_date__without_num_samples__station_number__without_NaN.copy()

In [21]:
# Now, we can save all bool-data as integer

weather_data['fog'] = weather_data['fog'].astype('int')
weather_data['rain'] = weather_data['rain'].astype('int')
weather_data['snow'] = weather_data['snow'].astype('int')
weather_data['hail'] = weather_data['hail'].astype('int')
weather_data['thunder'] = weather_data['thunder'].astype('int')
weather_data['tornado'] = weather_data['tornado'].astype('int')

In [22]:
weather_data = weather_data.sort_values(by='date')

In [23]:
weather_data.head()

station_number  mean_temp  mean_dew_point  mean_sealevel_pressure  \
279591          725305  30.900000       21.799999             1028.699951   
335820          725316  40.299999       35.799999             1025.599976   
99194           725315  39.000000       34.299999             1026.000000   
266632          725330  38.400002       32.599998             1027.800049   
317660          725314  53.799999       52.400002             1023.799988   

        mean_visibility  mean_wind_speed  max_sustained_wind_speed  \
279591             10.0              8.6                      14.0   
335820              7.0              9.0                      13.0   
99194               6.5             10.6                      15.9   
266632              9.2              7.9                      15.9   
317660              4.3              4.1                       9.9   

        max_temperature  total_precipitation  fog  rain  snow  hail  thunder  \
279591        21.000000                 0.00    0     0     0     0        0   
335820        35.599998                 0.01    1     1     1     1        1   
99194         33.799999                 0.03    1     1     1     1        1   
266632        30.000000                 0.08    1     1     1     1        1   
317660        48.000000                 0.36    0     0     0     0        0   

        tornado       date  
279591        0 2005-01-01  
335820        1 2005-01-01  
99194         1 2005-01-01  
266632        1 2005-01-01  
317660        0 2005-01-01

In [24]:
# The cleaned data
weather_data_cleaned = weather_data.copy()

In [25]:
weather_data_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17056 entries, 279591 to 1339
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   station_number            17056 non-null  int64         
 1   mean_temp                 17056 non-null  float64       
 2   mean_dew_point            17056 non-null  float64       
 3   mean_sealevel_pressure    17056 non-null  float64       
 4   mean_visibility           17056 non-null  float64       
 5   mean_wind_speed           17056 non-null  float64       
 6   max_sustained_wind_speed  17056 non-null  float64       
 7   max_temperature           17056 non-null  float64       
 8   total_precipitation       17056 non-null  float64       
 9   fog                       17056 non-null  int64         
 10  rain                      17056 non-null  int64         
 11  snow                      17056 non-null  int64         
 12  hail          

In [26]:
weather_data_cleaned.reset_index(drop=True, inplace=True)

In [27]:
weather_data_cleaned

station_number  mean_temp  mean_dew_point  mean_sealevel_pressure  \
0              725305  30.900000       21.799999             1028.699951   
1              725316  40.299999       35.799999             1025.599976   
2              725315  39.000000       34.299999             1026.000000   
3              725330  38.400002       32.599998             1027.800049   
4              725314  53.799999       52.400002             1023.799988   
...               ...        ...             ...                     ...   
17051          725315  64.900002       49.500000             1019.900024   
17052          725317  70.099998       53.000000             1020.200012   
17053          725314  67.900002       52.599998             1021.299988   
17054          725305  66.000000       50.200001             1017.000000   
17055          725316  67.099998       52.099998             1020.200012   

       mean_visibility  mean_wind_speed  max_sustained_wind_speed  \
0                 10.0              8.6                      14.0   
1                  7.0              9.0                      13.0   
2                  6.5             10.6                      15.9   
3                  9.2              7.9                      15.9   
4                  4.3              4.1                       9.9   
...                ...              ...                       ...   
17051             10.0              8.6                       9.9   
17052             10.0              6.6                       8.0   
17053             10.0              3.4                       6.0   
17054             10.0              6.8                       9.9   
17055             10.0              8.5                       9.9   

       max_temperature  total_precipitation  fog  rain  snow  hail  thunder  \
0            21.000000                 0.00    0     0     0     0        0   
1            35.599998                 0.01    1     1     1     1        1   
2            33.799999                 0.03    1     1     1     1        1   
3            30.000000                 0.08    1     1     1     1        1   
4            48.000000                 0.36    0     0     0     0        0   
...                ...                  ...  ...   ...   ...   ...      ...   
17051        61.000000                 0.00    0     0     0     0        0   
17052        68.000000                 0.00    0     0     0     0        0   
17053        64.900002                 0.00    0     0     0     0        0   
17054        63.000000                 0.00    0     0     0     0        0   
17055        64.000000                 0.00    0     0     0     0        0   

       tornado       date  
0            0 2005-01-01  
1            1 2005-01-01  
2            1 2005-01-01  
3            1 2005-01-01  
4            0 2005-01-01  
...        ...        ...  
17051        0 2010-04-16  
17052        0 2010-04-16  
17053        0 2010-04-16  
17054        0 2010-04-16  
17055        0 2010-04-16  

[17056 rows x 16 columns]

### 3. Task
Now it is time to split the data, into a training, evaluation and test set. As a reminder, the date we are trying to predict snow fall for is the following, and hence should constitute your test set.

In [28]:
import datetime, timedelta
# +3 leap years, -1 - tomorrow 
wanted_date = str(datetime.datetime.today()- datetime.timedelta(days=11*365 + 3 - 1)).split(' ')[0]
print("Wanted date: " + wanted_date + ".")

Wanted date: 2010-05-25.


In [29]:
all_dates = weather_data_cleaned["date"].dt.date

min_date = str(all_dates.min())
max_date = str(all_dates.max())
print("The available dates are between " + min_date + " and " + max_date + ".")
print("So, the latest date is " + max_date + " and it is earlier as our wanted date " + wanted_date + ".")

The available dates are between 2005-01-01 and 2010-04-16.
So, the latest date is 2010-04-16 and it is earlier as our wanted date 2010-05-25.


This means that we do not have to pay attention to which data is saved in which set.

My idea is to make the snowfall forecast "overnight". So, we have all the data from day (n) and want to know whether it will snow on the (n + 1)-th day. For this I have to expand the data with the column "snow tomorrow", where it is saved whether it will snow the next day. Then I can randomly split the data into subsets.

In [30]:
# Create new column "yesterday"
yesterday = pd.to_datetime(weather_data_cleaned.date) - datetime.timedelta(days=1)
weather_data_cleaned['yesterday'] = yesterday

In [31]:
# Create new dataframes to save dates and stations with or without snowfall
day_station = weather_data_cleaned.filter(['yesterday', 'station_number', 'snow'])
day_station_snow = day_station.loc[day_station['snow'] == 1]
day_station_no_snow = day_station.loc[day_station['snow'] == 0]

In [32]:
# Merge data: now we have the days, after then it snowed / not snowed
weather_data_cleaned_snow = weather_data_cleaned[weather_data_cleaned.set_index(['date','station_number']).index.isin(day_station_snow.set_index(['yesterday','station_number']).index)]
weather_data_cleaned_no_snow = weather_data_cleaned[weather_data_cleaned.set_index(['date','station_number']).index.isin(day_station_no_snow.set_index(['yesterday','station_number']).index)]

In [33]:
weather_data_cleaned_snow['snow_tomorrow'] = 1
weather_data_cleaned_no_snow['snow_tomorrow'] = 0

<ipython-input-33-840fef6c76df>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  weather_data_cleaned_snow['snow_tomorrow'] = 1
<ipython-input-33-840fef6c76df>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  weather_data_cleaned_no_snow['snow_tomorrow'] = 0


In [34]:
# Merge data
weather_data_cleaned_all = weather_data_cleaned_snow.append(weather_data_cleaned_no_snow, ignore_index=True)

In [35]:
weather_data_cleaned_all = weather_data_cleaned_all.drop([
    'yesterday',
], axis = 1)

In [36]:
weather_data_cleaned_all.head()

station_number  mean_temp  mean_dew_point  mean_sealevel_pressure  \
0          725305  30.900000       21.799999             1028.699951   
1          725316  40.299999       35.799999             1025.599976   
2          725315  39.000000       34.299999             1026.000000   
3          725330  38.400002       32.599998             1027.800049   
4          725300  32.099998       21.200001             1027.800049   

   mean_visibility  mean_wind_speed  max_sustained_wind_speed  \
0             10.0              8.6                      14.0   
1              7.0              9.0                      13.0   
2              6.5             10.6                      15.9   
3              9.2              7.9                      15.9   
4              9.9              8.7                      14.0   

   max_temperature  total_precipitation  fog  rain  snow  hail  thunder  \
0        21.000000                 0.00    0     0     0     0        0   
1        35.599998                 0.01    1     1     1     1        1   
2        33.799999                 0.03    1     1     1     1        1   
3        30.000000                 0.08    1     1     1     1        1   
4        24.799999                 0.00    0     0     0     0        0   

   tornado       date  snow_tomorrow  
0        0 2005-01-01              1  
1        1 2005-01-01              1  
2        1 2005-01-01              1  
3        1 2005-01-01              1  
4        0 2005-01-01              1

In [37]:
weather_data_cleaned_all['snow_tomorrow'].value_counts()

0    14631
1     2175
Name: snow_tomorrow, dtype: int64

As we can see, there are significantly fewer days with snowfall than without. It is important that this ratio is preserved after dividing into subsets. 

In [38]:
def split_train_eval_test(df, frac_train=0.65, frac_eval=0.1, frac_test=0.25):
    '''
    Splits a Pandas dataframe into training, evaluation and test sets.
    Each subset size can be set.
    Each set has the same relative frequency of the values in the column 'snow_tomorrow'.
    '''
    if frac_train + frac_eval + frac_test != 1.0:
        raise ValueError('fractions %f, %f, %f do not add up to 1.0' % \
                         (frac_train, frac_eval, frac_test))

    stratify_col = 'snow_tomorrow'
    
    X = df # All columns
    #X = df.drop([stratify_col], axis=1) # All columns, except 'snow_tomorrow'
    y = df[[stratify_col]] # Dataframe of just the 'snow_tomorrow' column

    # Split input dataframe into train and eval_test dataframes.
    train, eval_test, y_train, y_eval_test = train_test_split(X,
                                                          y,
                                                          stratify=y,
                                                          test_size=(1.0 - frac_train))

    # Split the eval_test dataframe into eval and test dataframes.
    relative_frac_test = frac_test / (frac_eval + frac_test)
    _eval, test, y_eval, y_test = train_test_split(eval_test,
                                                      y_eval_test,
                                                      stratify=y_eval_test,
                                                      test_size=relative_frac_test)
    
    X_train = train.drop([stratify_col], axis=1) # All columns, except 'snow_tomorrow'
    X_eval = _eval.drop([stratify_col], axis=1) # All columns, except 'snow_tomorrow'
    X_test = test.drop([stratify_col], axis=1) # All columns, except 'snow_tomorrow'
    
    assert len(df) == len(X_train) + len(X_eval) + len(X_test)

    return train, _eval, test

In [39]:
train, _eval, test = split_train_eval_test(weather_data_cleaned_all)

In [40]:
# Save data as .csv to use it in part 2

weather_data_cleaned_all.to_csv('weather_data_cleaned_all.csv')
train.to_csv('train.csv')
_eval.to_csv('_eval.csv')
test.to_csv('test.csv')